### PR TITLE
chore: upgrade datafusion and arrow-rs

### PR DESCRIPTION
**Pull request by wjones127**

### CODE REVIEW

File: rust/lance/src/io/reader.rs [hunk 1]
```diff
@@ -1528,22 +1529,22 @@ mod tests {
 
         let arr = read_array_w_params(&reader, &schema.fields[1], ReadBatchParams::RangeFull).await;
         assert_eq!(100, arr.len());
-        assert_eq!(100, arr.null_count());
+        assert_eq!(arr.data_type(), &DataType::Null);
```
Comment (wjones127) [inline rust/lance/src/io/reader.rs:1532]: `null_count()` no longer returns the logical null count, just the number of "physical nulls". https://github.com/apache/arrow-rs/pull/4691

File: rust/Cargo.toml [hunk 2]
```diff
@@ -42,23 +42,24 @@ lance-linalg = { version = "=0.8.3", path = "./lance-linalg" }
...
-arrow-array = "43.0"
-arrow-buffer = "43.0"
-arrow-cast = "43.0"
-arrow-data = "43.0"
-arrow-ord = "43.0"
-arrow-row = "43.0"
-arrow-schema = "43.0"
-arrow-select = "43.0"
+arrow = { version = "46.0.0", optional = false }
+arrow-arith = "46.0"
```
Comment (eddyxu) [inline rust/Cargo.toml:46]: IIRC, we are wait for arrow 47 for a few fixes? And datafusion 32 is being voted at the moment, right?

Comment (wjones127) [reply rust/Cargo.toml:46]: Why wait?

Comment (eddyxu) [reply rust/Cargo.toml:46]: Do we need to bump `datafusion` again in 3-5 days to include fixes?

Comment (wjones127) [reply rust/Cargo.toml:46]: 🤷 yes, we need to upgrade for the next version. IMO we should upgrade on every datafusion release.

**Pull request closed without merging**